### PR TITLE
[3/5] light client events

### PR DESCRIPTION
**Pull request by qinlz2**
**What type of PR is this?**

Feature

** What does this PR do? Why is it needed?**

Implements the Altair spec logic defined in this version:

https://github.com/ethereum/consensus-specs/tree/208da34ac4e75337baf79adebf036ab595e39f15/specs/altair/light-client

It also implements the relevant endpoints specified at https://github.com/ethereum/beacon-APIs

Specifically, this PR adding light client event sub/pub for `optimistic update` and `finalized update`

https://github.com/prysmaticlabs/prysm/pull/11815

**Which issues(s) does this PR fix?**

Partially addresses https://github.com/prysmaticlabs/prysm/issues/11571

Other notes for review

This PR is a follow up to https://github.com/prysmaticlabs/prysm/pull/11815 which was too big to be reviewed. This is the third PR in a series of 5 that I will be submitting sequentially.

The heavy lifting here was done by [@qinlz2](https://github.com/qinlz2) and [@7AC](https://github.com/7AC), so kudos to them.

### CODE REVIEW

File: beacon-chain/rpc/apimiddleware/structs.go [hunk 1]
```diff
@@ -17,6 +17,30 @@ type WeakSubjectivityResponse struct {
 	} `json:"data"`
 }
 
+type LightClientFinalityUpdateResponseJson struct {
```
Comment (qinlz2) [inline beacon-chain/rpc/apimiddleware/structs.go:20]: We understand there is an ongoing efforts to remove `apimiddleware` structs.  

Unfortunately, our original implementation was written before that.   

Please help to give some guidance (or reference PR), and we will be happy to refactor.

**Comment by qinlz2** [comment]
@rkapka Thanks for helping us with previous 2 PRs.  
This is the 3rd one in the series.   Please help to review it at your convenient time.  Thanks!

**Comment by rkapka** [comment]
Hi @qinlz2 , thanks a lot for this! We recently completed our move to HTTP for all Beacon API endpoints, including events. Events endpoint was moved here: https://github.com/prysmaticlabs/prysm/pull/13207. We also removed the API Middleware entirely in https://github.com/prysmaticlabs/prysm/pull/13243. Can you rebase this PR so that it conforms to the current design?

**Comment by qinlz2** [comment]
> Hi @qinlz2 , thanks a lot for this! We recently completed our move to HTTP for all Beacon API endpoints, including events. Events endpoint was moved here: #13207. We also removed the API Middleware entirely in #13243. Can you rebase this PR so that it conforms to the current design?

Yes, working on it.  Thanks.

**Comment by qinlz2** [comment]
@rkapka We rebase to the latest `develop` branch and did the refactoring to be consistent style. 
Please help to take another look?   Thanks!

**Comment by rkapka** [comment]
@qinlz2 Thanks for this, I left review feedback

**Comment by rkapka** [comment]
Hi @qinlz2 , a gentle reminder to take a look at the review when you find some time.

File: beacon-chain/rpc/eth/events/structs.go [hunk 2]
```diff
@@ -92,3 +92,27 @@ type BlobSidecarEvent struct {
...
+
+type LightClientFinalityUpdateEvent struct {
+	Version string                     `json:"version" enum:"true"`
+	Data    *LightClientFinalityUpdate `json:"data"`
+}
+
+type LightClientFinalityUpdate struct {
+	AttestedHeader  *shared.BeaconBlockHeader `json:"attested_header"`
+	FinalizedHeader *shared.BeaconBlockHeader `json:"finalized_header"`
+	FinalityBranch  []string                  `json:"finality_branch"  hex:"true"`
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/events/structs.go:104]: Patched.

File: beacon-chain/rpc/eth/events/structs.go [hunk 3]
```diff
@@ -92,3 +92,27 @@ type BlobSidecarEvent struct {
 	KzgCommitment string `json:"kzg_commitment"`
 	VersionedHash string `json:"versioned_hash"`
 }
+
+type LightClientFinalityUpdateEvent struct {
+	Version string                     `json:"version" enum:"true"`
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/rpc/eth/events/structs.go:97]: Patched.

File: beacon-chain/blockchain/process_block.go [hunk 4]
```diff
@@ -41,6 +43,85 @@ const depositDeadline = 20 * time.Second
...
+	}
+
+	// Send event
+	return s.cfg.StateNotifier.StateFeed().Send(&feed.Event{
+		Type: statefeed.LightClientFinalityUpdate,
+		Data: result,
+	}), nil
+}
+
+// sendLightClientOptimisticUpdate sends a light client optimistic update notification of  to the state feed.
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/blockchain/process_block.go:92]: Patched.

File: beacon-chain/blockchain/process_block.go [hunk 5]
```diff
@@ -115,6 +196,16 @@ func (s *Service) postBlockProcess(ctx context.Context, signed interfaces.ReadOn
 		},
 	})
 
+	if _, err := s.sendLightClientOptimisticUpdate(ctx, signed, postState); err != nil {
+		log.WithError(err)
+	}
+
+	// Save finalized check point to db and more.
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/blockchain/process_block.go:203]: Patched.

File: beacon-chain/blockchain/process_block.go [hunk 6]
```diff
@@ -115,6 +196,16 @@ func (s *Service) postBlockProcess(ctx context.Context, signed interfaces.ReadOn
...
+	finalized := s.ForkChoicer().FinalizedCheckpoint()
+
+	// LightClientFinalityUpdate needs super majority
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/blockchain/process_block.go:206]: Patched.

File: beacon-chain/blockchain/process_block.go [hunk 7]
```diff
@@ -115,6 +196,16 @@ func (s *Service) postBlockProcess(ctx context.Context, signed interfaces.ReadOn
...
+	s.tryPublishLightClientFinalityUpdate(ctx, signed, finalized, postState)
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/blockchain/process_block.go]: Thanks.  Patched.

File: beacon-chain/blockchain/process_block.go [hunk 8]
```diff
@@ -144,6 +235,33 @@ func (s *Service) postBlockProcess(ctx context.Context, signed interfaces.ReadOn
...
+		return
+	}
+
+	if syncAggregate.SyncCommitteeBits.Count()*3 < config.SyncCommitteeSize*2 {
+		return
+	}
+
+	_, err = s.sendLightClientFinalityUpdate(ctx, signed, postState)
+	if err != nil {
+		log.WithError(err)
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/blockchain/process_block.go:262]: Patched.

File: beacon-chain/blockchain/process_block.go [hunk 9]
```diff
@@ -115,6 +196,16 @@ func (s *Service) postBlockProcess(ctx context.Context, signed interfaces.ReadOn
 		},
 	})
 
+	if _, err := s.sendLightClientOptimisticUpdate(ctx, signed, postState); err != nil {
+		log.WithError(err)
```
[the remark this answers is not in the document]
Comment (qinlz2) [reply beacon-chain/blockchain/process_block.go]: Patched.

**Comment by qinlz2** [comment]
> Hi @qinlz2 , a gentle reminder to take a look at the review when you find some time.

Thanks @rkapka .  Sorry was out of town. 
I have addressed all the comments and rebase to the latest `develop` branch.   

Happy holidays!

File: config/features/config.go [hunk 10]
```diff
@@ -41,6 +42,7 @@ type Flags struct {
 	WriteSSZStateTransitions            bool // WriteSSZStateTransitions to tmp directory.
 	EnablePeerScorer                    bool // EnablePeerScorer enables experimental peer scoring in p2p.
 	DisableReorgLateBlocks              bool // DisableReorgLateBlocks disables reorgs of late blocks.
+	EnableLightClientEvents             bool // EnableLightClientEvents enables light client events to be emitted by the beacon node.
```
Comment (rkapka) [inline config/features/config.go:45]: Can you rename this to `EnableLightClient` please? We will probably want to use this flag in upcoming PRs too for things other than events e.g. saving to the DB. Therefore a more general name will be better suited.

**Comment by rkapka** [comment]
Hey @qinlz2 , I left one last comment. I would like to test the functionality after this PR is merged. Are you a member of Prysm Discord server, so that I can reach out to you in case of questions (if you don't mind, of course)?

File: config/features/config.go (hunk 10 above)
Comment (qinlz2) [reply config/features/config.go]: Patched.

**Comment by qinlz2** [comment]
> Hey @qinlz2 , I left one last comment. I would like to test the functionality after this PR is merged. Are you a member of Prysm Discord server, so that I can reach out to you in case of questions (if you don't mind, of course)?

Thanks @rkapka 
* I patched based on your last comment. 
* I am in Prysm Discord server.   Feel free to ping me there.

**Comment by rkapka** [comment]
Hey @qinlz2 , I would like to approve this but:
- the branch accumulated some conflicts
- `TestLightClientHandler_GetLightClientUpdatesByRange` fails

**Comment by qinlz2** [comment]
> Hey @qinlz2 , I would like to approve this but:
> 
> * the branch accumulated some conflicts
> * `TestLightClientHandler_GetLightClientUpdatesByRange` fails

Thanks, @rkapka.  Sorry I missed this. 

* Rebased to the latest develop branch
* Fixed the broken unit tests.

**Review by rkapka — APPROVED** [review]

**Comment by rkapka** [comment]
Thanks @qinlz2 ! We're almost set - can you rebase one more time please?

**Comment by Inspector-Butters** [comment]
Hey @rkapka,
I'm going to try to continue and hopefully finish this work for my EPF project. 
can you please confirm that this [3/5] is the latest changes that have been implemented and merged?
Thank you. 

Also I could not find you in the Eth R&D discord server, please share some way to contact you or ping me if possible.

**Pull request merged**